lib/src/revset_graph.rs

```diff
@@ -111,29 +115,39 @@ impl Iterator for ReverseRevsetGraphIterator {
 ///
 /// [Git]: https://github.blog/2022-08-30-gits-database-internals-ii-commit-history-queries/#topological-sorting
 #[derive(Clone, Debug)]
-pub struct TopoGroupedRevsetGraphIterator<I> {
+pub struct TopoGroupedRevsetGraphIterator<N, I> {
     input_iter: I,
     /// Graph nodes read from the input iterator but not yet emitted.
-    nodes: HashMap<CommitId, TopoGroupedGraphNode>,
+    nodes: HashMap<N, TopoGroupedGraphNode<N>>,
     /// Stack of graph nodes to be emitted.
-    emittable_ids: Vec<CommitId>,
+    emittable_ids: Vec<N>,
     /// List of new head nodes found while processing unpopulated nodes.
-    new_head_ids: VecDeque<CommitId>,
+    new_head_ids: VecDeque<N>,
     /// Set of nodes which may be ancestors of `new_head_ids`.
-    blocked_ids: HashSet<CommitId>,
+    blocked_ids: HashSet<N>,
 }
 
-#[derive(Clone, Debug, Default)]
-struct TopoGroupedGraphNode {
+#[derive(Clone, Debug)]
+struct TopoGroupedGraphNode<N> {
     /// Graph nodes which must be emitted before.
-    child_ids: HashSet<CommitId>,
+    child_ids: HashSet<N>,
     /// Graph edges to parent nodes. `None` until this node is populated.
-    edges: Option<Vec<RevsetGraphEdge>>,
+    edges: Option<Vec<RevsetGraphEdge<N>>>,
 }
 
-impl<I> TopoGroupedRevsetGraphIterator<I>
+impl<N> Default for TopoGroupedGraphNode<N> {
+    fn default() -> Self {
+        Self {
+            child_ids: Default::default(),
+            edges: None,
+        }
+    }
+}
```

Comment:
Is Rust not able to derive this? Does it think that `N` must implement `Default` for it to work or what's the issue?

Reply:
Yup, that's a classic one, derive only works for Option<T> if T is Default, you guessed it